### PR TITLE
LPS-198464 OOTB SXP Element features a Site Selector so the user can browse and select the sites to limit the search to without the need look-up for and copy-paste the groupIds (Site IDs)

### PR DESCRIPTION
**Pull request by felipelorenz**
https://liferay.atlassian.net/browse/LPS-198464

### CODE REVIEW

**Comment by felipelorenz** [comment]
ci:test:sf

File: modules/dxp/apps/search-experiences/search-experiences-web/src/main/java/com/liferay/search/experiences/web/internal/blueprint/admin/portlet/action/GetSitesMVCResourceCommand.java [hunk 1]
```diff
@@ -122,10 +122,29 @@ protected JSONObject getSitesJSONObject(
...
+		int page = ParamUtil.getInteger(resourceRequest, "page");
+
+		if (page > lastPage) {
+			page = lastPage;
+		}
+
+		int endPosition = pageSize * page;
+
+		int startPosition = endPosition - pageSize;
+
```
Comment (joshuacords) [inline modules/dxp/apps/search-experiences/search-experiences-web/src/main/java/com/liferay/search/experiences/web/internal/blueprint/admin/portlet/action/GetSitesMVCResourceCommand.java:139]: Naming would be better as:
pageSize -> delta
totalCount -> total
endPostition -> end
startPosition -> start

**Comment by joshuacords** [comment]
Hello @felipelorenz, I was trying to run your test using `gw test --tests .*GetSites.*` but I just get an error 

> Execution failed for task ':dxp:apps:search-experiences:search-experiences-service:test'.
> > No tests found for given includes: [**/*Test.class](include rules) [.*GetSites.*](--tests filter)

Does this work on your machine?

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	private String _fixElementDefinition(String elementDefinition)
+		throws Exception {
+
+		if (Validator.isBlank(elementDefinition)) {
+			return elementDefinition;
+		}
+
+		JSONObject sxpElementJSONObject = JSONFactoryUtil.createJSONObject(
```
Comment (joshuacords) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java:37]: Wrap `JSONFactoryUtil.createJSONObject()` in a try-catch.

**Review by joshuacords — CHANGES_REQUESTED** [review]
Rename the variables, let me know if and how to run the test, and wrap createJSON in a try-catch.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_1/SXPBlueprintUpgradeProcess.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+				else if (defaultValueJSONObject != null) {
+					fieldJSONObject.put("defaultValue", StringPool.BLANK);
+				}
+			}
+		}
+
+		return sxpElementJSONObject.toString();
+	}
+
+	private String _fixElementInstance(String elementInstances)
```
Comment (peerkar) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_1/SXPBlueprintUpgradeProcess.java:91]: If possible, I'd consider renaming it to something more descriptive. "fixXYZ" doesn't tell much.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+public class SXPBlueprintUpgradeProcess extends UpgradeProcess {
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		_upgradeSXPElement();
+
+		_upgradeSXPBlueprint();
+	}
+
+	private String _fixElementDefinition(String elementDefinition)
```
Comment (peerkar) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java:30]: If possible, I'd consider renaming it to something more descriptive. "fixXYZ" doesn't tell much.

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_1/SXPBlueprintUpgradeProcess.java (hunk 3 above)
Comment (gustavolimav) [reply modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_1/SXPBlueprintUpgradeProcess.java:91]: https://github.com/brianchandotcom/liferay-portal/commit/e871b34f75b1de96155b9ec9b3bf856a468e8f24

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java (hunk 4 above)
Comment (gustavolimav) [reply modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java:30]: https://github.com/brianchandotcom/liferay-portal/commit/e871b34f75b1de96155b9ec9b3bf856a468e8f24

Comment (peerkar) [reply modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java:30]: @gustavolimav this is not about getting the PR accepted, but making readable and understandable code. For the others and for us, as well. Consider each and every upgrade process having a method `fixElement()` and you having to quickly figure out what each of these actually do. Compare then the same, when each of those methods are actually saying what they do, like `fixHelpTextLocalizations()`, `addDefaultTitles()` etc.

Comment (gustavolimav) [reply modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/internal/upgrade/v3_1_2/SXPBlueprintUpgradeProcess.java:30]: Yeah, I agree 100% just put here as reference and so Felipe can change it in my code too

File: modules/dxp/apps/search-experiences/search-experiences-web/src/main/java/com/liferay/search/experiences/web/internal/blueprint/admin/portlet/action/GetSitesMVCResourceCommand.java (hunk 1 above)
Comment (gustavolimav) [reply modules/dxp/apps/search-experiences/search-experiences-web/src/main/java/com/liferay/search/experiences/web/internal/blueprint/admin/portlet/action/GetSitesMVCResourceCommand.java:139]: pageSize, page, and totalCount seem to be standard in the headless APIs

**Comment by gustavolimav** [comment]
> Hello @felipelorenz, I was trying to run your test using `gw test --tests .*GetSites.*` but I just get an error
> 
> > Execution failed for task ':dxp:apps:search-experiences:search-experiences-service:test'.
> > > No tests found for given includes: [**/*Test.class](include rules) [._GetSites._](--tests filter)
> 
> Does this work on your machine?

I was able to run using `gw test --tests "GetSitesMVCResourceCommandTest"`

@joshuacords

**Comment by joshuacords** [comment]
Ah - I was trying to run the test in search-experiences-service instead of search-experiences-web

**Comment by gustavolimav** [comment]
Resent [here](https://github.com/liferay-search/liferay-portal/pull/1202)

**Pull request closed without merging**